### PR TITLE
feat(ProductStocks): build data models

### LINKED ISSUE #40: Product Stocks - build data model
See ERD on readme

### PR DESCRIPTION
**Pull request by mapra99**
Closes #40 
This PR adds these models
image

### CODE REVIEW

File: app/models/topping.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+  PRICE_CHANGE_REGEXP = /\A(\+|-)(\d+\.?\d*)\z/
+
+  belongs_to :product
+  has_many :stock_toppings, dependent: :destroy
+  has_many :stocks, through: :stock_toppings
+
+  validates :key, presence: true, uniqueness: { scope: %i[value product_id] }
+  validates :value, presence: true
+  validates :price_change,
+            format: { with: PRICE_CHANGE_REGEXP, message: 'must follow the price change format (e.g. +120.2)' }
```
Comment (ImMPrada) [inline app/models/topping.rb:11]: This is to increase or decrease the base value at any time, right?

File: db/migrate/20220704222524_create_stocks.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class CreateStocks < ActiveRecord::Migration[6.1]
+  def change
+    create_table :stocks do |t|
+      t.references :product, null: false, foreign_key: true
```
Comment (ImMPrada) [inline db/migrate/20220704222524_create_stocks.rb:4]: product 1:n stock?

File: db/migrate/20220704223128_create_toppings.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+class CreateToppings < ActiveRecord::Migration[6.1]
+  def change
+    create_table :toppings do |t|
+      t.references :product, null: false, foreign_key: true
```
Comment (ImMPrada) [inline db/migrate/20220704223128_create_toppings.rb:4]: topping 1:n product?

**Review by ImMPrada — COMMENTED** [review]
models review

**Review by ImMPrada — APPROVED** [review]
I learn more than what I can check xD

**Comment by ImMPrada** [comment]
PR #40 says ERD is at README, but I didn't see it

File: app/models/topping.rb (hunk 1 above)
Comment (mapra99) [reply app/models/topping.rb:11]: yeah, the idea is that a `product` would have a `base_price`, and the toppings (for example, color or size variations) would be able to modify that price by adding or substracting to the `base_price`

File: db/migrate/20220704222524_create_stocks.rb (hunk 2 above)
Comment (mapra99) [reply db/migrate/20220704222524_create_stocks.rb:4]: yeah, a product has many stocks, one stock per "toppings set", for example if we sell a headset:
product: Headset Beat X101
stocks:
  - 10 units of color black and size M
  - 10 units of color white and size M
  - 5 units of color black and size S
  - 2 units of color white and size S

File: db/migrate/20220704223128_create_toppings.rb (hunk 3 above)
Comment (mapra99) [reply db/migrate/20220704223128_create_toppings.rb:4]: yeah, all the toppings of a product

**Comment by mapra99** [comment]
> PR https://github.com/mapra99/audiophile/issues/40 says ERD is at README, but I didn't see it

@ImMPrada  That's an issue, not a PR, and yeah 🤦  I was referring to the project README, not the repo README. It's not very visible though, I should add it on the repo
https://github.com/users/mapra99/projects/1/

**Comment by mapra99** [comment]
For reference, I'll work on adding the ERD on #53

**Pull request merged**